### PR TITLE
setup commit message linting

### LINKED ISSUE #4: Setting up Commitlint + Rules
Quoting Bultz from Discord who summed up this nicely. This is from the `#wiki-working-group` in the [Battlesnake Discord](https://discord.gg/gPYEZEQUyp)

> So for commits, are there a general consensus of:
> - type and scope = error 
> - rest = warning
> 
> type and scope are in the conventional commits, so if they use that, doesn't really look like it's someone first time. So I would say it's fair. 
> 
> Not sure if the only thing to say we enforce is the max 50 char in the subject and warning on the rest or make the subject also a warning. 
> 
> it looks like we can change the warning and error messages, so we can make it more clear on the WHY, instead of "this is a warning", "this is a error"

Links:
https://github.com/conventional-changelog/commitlint

### PR DESCRIPTION
**Pull request by DaBultz**
This closes #4 

Commitlint did not allow overriding the error messages, however it allowed custom rules (with custom errors) which is the approach I chose. I only implemented 4 rules for now.

it's recommended that commitlint runs on the commit-msg git hook, to make sure this happened husky was added to make sure that the hooks are properly configured and automatically after `npm install`. 

### `docs/no-type` & `docs/no-scope`

These 2 rules disallows the use of the type and scope which is used in the conventional commit standard. these will create an error if they are found in the commit message 

### `docs/header-length` 

This rule will warn if the header is longer than 50 characters long

### `docs/body-blank-line`

This rule will warn if there is no blank line between the body and header

## Questions

1) Are we missing any rules?
2) How much info should each warning/error provide? (multiline, colors, bold, italic and more is possible)
3) Should we extend the contributing guide to include a section about git commit messages?

### CODE REVIEW

File: commitlint.config.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            "docs/no-type": (parsed) => {
+                return [
+                    parsed.type == null,
+                    `Type is not allowed, type found: '${parsed.type}'`
+                ];
+            },
+            "docs/no-scope": (parsed) => {
+                return [
+                    parsed.scope == null,
+                    `Scope is not allowed, sype found: '${parsed.scope}'`
```
Comment (coreyja) [inline commitlint.config.js]:
```suggestion
                    `Scope is not allowed, scope found: '${parsed.scope}'`
```

Comment (DaBultz) [reply commitlint.config.js]: Good catch, approved the suggestion :)

**Review by jlafayette — COMMENTED** [review]
Thanks for getting this setup!  Is there a step to install the commithook that I'm missing?  I'm seeing an error when I try an invalid message:

```
PS G:\code\docs-new> git commit -v -m "fix: edits"
G:\code\docs-new\node_modules\@commitlint\cli\lib\cli.js:114
        throw err;
        ^

[Error: ENOENT: no such file or directory, open 'G:\code\docs-new\${1}'] {
  errno: -4058,
  code: 'ENOENT',
  syscall: 'open',
  path: 'G:\\code\\docs-new\\${1}'
}

Node.js v18.4.0
husky - commit-msg hook exited with code 1 (error)
```

If there is, then that might a good thing to add to the README or CONTRIBUTING docs.

All I did was checkout the branch and run `npm install`

**Comment by DaBultz** [comment]
> Thanks for getting this setup! Is there a step to install the commithook that I'm missing? I'm seeing an error when I try an invalid message:
> 
> ```
> PS G:\code\docs-new> git commit -v -m "fix: edits"
> G:\code\docs-new\node_modules\@commitlint\cli\lib\cli.js:114
>         throw err;
>         ^
> 
> [Error: ENOENT: no such file or directory, open 'G:\code\docs-new\${1}'] {
>   errno: -4058,
>   code: 'ENOENT',
>   syscall: 'open',
>   path: 'G:\\code\\docs-new\\${1}'
> }
> 
> Node.js v18.4.0
> husky - commit-msg hook exited with code 1 (error)
> ```
> 
> If there is, then that might a good thing to add to the README or CONTRIBUTING docs.
> 
> All I did was checkout the branch and run `npm install`



> Thanks for getting this setup! Is there a step to install the commithook that I'm missing? I'm seeing an error when I try an invalid message:
> 
> ```
> PS G:\code\docs-new> git commit -v -m "fix: edits"
> G:\code\docs-new\node_modules\@commitlint\cli\lib\cli.js:114
>         throw err;
>         ^
> 
> [Error: ENOENT: no such file or directory, open 'G:\code\docs-new\${1}'] {
>   errno: -4058,
>   code: 'ENOENT',
>   syscall: 'open',
>   path: 'G:\\code\\docs-new\\${1}'
> }
> 
> Node.js v18.4.0
> husky - commit-msg hook exited with code 1 (error)
> ```
> 
> If there is, then that might a good thing to add to the README or CONTRIBUTING docs.
> 
> All I did was checkout the branch and run `npm install`

Should be fixed, forgot that windows handles `$1` differently.

**Review by jlafayette — APPROVED** [review]
Thanks for fixing that!  It seems to be working on Windows now

File: commitlint.config.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            },
+            "docs/header-length": (parsed) => {
+                return [
+                    parsed.header.length <= 50,
+                    "Keep the header under 50 characters",
+                ]
+            },
+            "docs/body-blank-line": (parsed) => {
+                // get the first line in the body
+                const leading = parsed.raw.split(/(?:\r?\n)/).slice(1);
```
Comment (coreyja) [inline commitlint.config.js]: Does this `slice` want to be an array index?

```suggestion
                const leading = parsed.raw.split(/(?:\r?\n)/)[1];
```

This rule wasn't working for me locally and I think slice is returning an array here, and won't be equal to the empty string below?

Comment (DaBultz) [reply commitlint.config.js]: What commit message did you try? just so i can double check locally.

Comment (coreyja) [reply commitlint.config.js]:
```
test
test
test
```

lol

**Review by coreyja — APPROVED** [review]
💯

File: commitlint.config.js (hunk 2 above)
Comment (DaBultz) [reply commitlint.config.js]: Try and see if the newest push fixes it.

Comment (coreyja) [reply commitlint.config.js]: Yup works great!

**Comment by DaBultz** [comment]
made my branch up to date, so it's possible to merge it.

**Pull request merged**